duolingo.py

```diff
@@ -28,20 +28,84 @@ class DuolingoException(Exception):
 class Duolingo(object):
     USER_AGENT = "Python Duolingo API/{}".format(__version__)
 
-    def __init__(self, username, password):
-        self.username = username
-        self.password = password
-        self.user_url = "https://duolingo.com/users/%s" % self.username
+    _url_base = "https://{host}{base_path}"
+    _url_user_no_auth = "/users?username={username}"
+    _url_login = "/login?fields="
+
+    def __init__(self, username="", password="", host="www.duolingo.com", base_path="/2017-06-30"):
+        self.host = host
+        self.base_path = base_path
+        self.base_url = self._url_base.format(host=self.host, base_path=self.base_path)
+
         self.session = requests.Session()
         self.leader_data = None
-        self.jwt = None
+        self.jwt = None        
+
+        if username:
+            self.username = username
+            if password:
+                self.password = password
+                self._login()
+            self._get_user_data(username)
+        self.voice_url_dict = None
 
-        if password:
-            self._login()
+    def _get_user_data(self, username):
+        """Gets user data with _get_data if logged in
+            and _get_user_data_no_auth otherwise
+        
+        Arguments:
+            username {str} -- username on Duolingo
+        """        
+        self.username = username
+        self.user_url = "https://duolingo.com/users/%s" % self.username # TODO: should be removed
+
+        data = self._get_data() if self.jwt else self._get_user_data_no_auth(username)
+        self.user_data = Struct(**data)
+
+    def _get_user_data_no_auth(self, username):
+        """Gets user data without authentication
+        
+        Arguments:
+            username {str} -- username on Duolingo
+        
+        Returns:
+            dict -- user data
+        """        
+        url = self.base_url + self._url_user_no_auth.format(username=username)
+        get = self._make_req(url).json()
+        if not "users" in get:
+            raise DuolingoException("users field not found in response")
+        if len(get["users"]) == 0:
+            raise DuolingoException("No users in response")
+        return get["users"][0]
+
+    def login(self, username, password):
+        """Authencicate
+        
+        Arguments:
+            username {str} -- username on Duolingo
+            password {str} -- user password
+        
+        Raises:
+            DuolingoException: Login failed
+        
+        Returns:
+            bool -- Login success if True
+        """        
+        str
```

Comment:
Left a loose str in here

Reply:
Thanks, sorry for that. I'll check how to add commit to the PR

Reply:
They should just appear on the PR when you push commits to the branch, which is convenient